src/System.Runtime.Extensions/src/System/IO/Path.Unix.cs

```diff
@@ -55,82 +57,21 @@ public static string GetFullPath(string path)
-                    // e.g. "parent//child" => "parent/child"
-                    if (PathInternal.IsDirectorySeparator(path[i + 1]))
-                    {
-                        continue;
-                    }
-
-                    // Skip this character and the next if it's referring to the current directory,
-                    // e.g. "parent/./child" =? "parent/child"
-                    if ((i + 2 == path.Length || PathInternal.IsDirectorySeparator(path[i + 2])) &&
-                        path[i + 1] == '.')
-                    {
-                        i++;
-                        continue;
-                    }
-
-                    // Skip this character and the next two if it's referring to the parent directory,
-                    // e.g. "parent/child/../grandchild" => "parent/grandchild"
-                    if (i + 2 < path.Length &&
-                        (i + 3 == path.Length || PathInternal.IsDirectorySeparator(path[i + 3])) &&
-                        path[i + 1] == '.' && path[i + 2] == '.')
-                    {
-                        // Unwind back to the last slash (and if there isn't one, clear out everything).
-                        int s;
-                        for (s = sb.Length - 1; s >= 0; s--)
-                        {
-                            if (PathInternal.IsDirectorySeparator(sb[s]))
-                            {
-                                sb.Length = s;
-                                break;
-                            }
-                        }
-                        if (s < 0)
-                            sb.Length = 0;
-
-                        i += 2;
-                        continue;
-                    }
-                }
+            string collapsedString = RemoveRelativeSegments(path);
 
-                if (++componentCharCount > MaxComponentLength)
-                {
-                    throw new PathTooLongException(SR.IO_PathTooLong);
-                }
-                sb.Append(c);
-            }
-
-            Debug.Assert(sb.Length < path.Length || sb.ToString() == path,
+            Debug.Assert(collapsedString.Length < path.Length || collapsedString.ToString() == path,
                 "Either we've removed characters, or the string should be unmodified from the input path.");
 
-            if (sb.Length > MaxPath)
+            if (collapsedString.Length > MaxPath)
             {
                 throw new PathTooLongException(SR.IO_PathTooLong);
             }
 
             string result =
-                sb.Length == 0 ? (fullCheck ? DirectorySeparatorCharAsString : string.Empty) :
-                sb.Length == path.Length ? path :
-                sb.ToString();
-            StringBuilderCache.Release(sb);
+                collapsedString.Length == 0 ? (fullChecks ? DirectorySeparatorCharAsString : string.Empty) :
+                collapsedString.Length == path.Length ? path :
```

Comment:
Whoops- I'll move this to the helper and keep track of whether or not we actually modify the string.